sdk/log/batch.go

```diff
@@ -66,14 +107,96 @@ func (b *BatchingProcessor) Enabled(context.Context, Record) bool {
 
 // Shutdown flushes queued log records and shuts down the decorated exporter.
 func (b *BatchingProcessor) Shutdown(ctx context.Context) error {
-	// TODO (#5063): Implement.
-	return nil
+	wasShutdown := b.isShutdown.Swap(true)
+	if wasShutdown {
+		return nil
+	}
+
+	req := exportRequest{
+		Context: ctx,
+		Result:  make(chan error, 1), // Heap allocation.
+	}
+	b.stop <- req // Send to a buffered channel so that it eventually closes.
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case err := <-req.Result:
+		return err
+	}
 }
 
 // ForceFlush flushes queued log records and flushes the decorated exporter.
 func (b *BatchingProcessor) ForceFlush(ctx context.Context) error {
-	// TODO (#5063): Implement.
-	return nil
+	if b.isShutdown.Load() {
+		return nil
+	}
+
+	req := exportRequest{
+		Context: ctx,
+		Result:  make(chan error, 1), // Heap allocation.
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-b.done:
+		// The batching processor was concurrently shutdown.
+		return nil
+	case b.flush <- req:
+		return <-req.Result
+	}
+}
+
+func (b *BatchingProcessor) run() {
+	defer close(b.done)
+
+	ticker := time.NewTicker(b.exportInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			err := b.export(context.Background())
+			if err != nil {
+				otel.Handle(err)
+			}
+		case req := <-b.flush:
+			err := b.export(req.Context)
+			err = errors.Join(err, b.exporter.ForceFlush(req.Context))
+			req.Result <- err
+			ticker.Reset(b.exportInterval)
+		case req := <-b.stop:
+			err := b.export(req.Context)
+			err = errors.Join(err, b.exporter.Shutdown(req.Context))
+			req.Result <- err
+			return
+		}
+	}
+}
+
+func (b *BatchingProcessor) export(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, b.exportTimeout) // 5 heap allocations.
+	defer cancel()
+
+	b.mu.Lock()
+	if len(b.queue) == 0 {
+		// Nothing to export
+		b.mu.Unlock()
+		return nil
+	}
+	if len(b.queue) > b.exportMaxBatchSize {
+		b.batch = append(b.batch, b.queue[:b.exportMaxBatchSize]...)
+		b.queue = b.queue[b.exportMaxBatchSize:]
+	} else {
+		b.batch = append(b.batch, b.queue...)
+		b.queue = b.queue[:0]
+	}
+	b.mu.Unlock()
+
+	// Doing export outside of the lock prevents deadlocks and improves efficiency.
+	err := b.exporter.Export(ctx, b.batch)
+	b.batch = b.batch[:0]
```

Comment:
Should we clear the batch if `Export` returned an error?
Or should it be

```suggestion
	if err == nil {
		b.batch = b.batch[:0]
	}
```

?